Fix issue #19484 - ACCEPT in (...)+ doesn't work as expected, plus other ACCEPT related bugs.

demerphq opened the pull request:
This PR fixes Issue #19484, along with some ancillary issues related to ACCEPT and look-behind. 

It also adds a bunch of tests.

It also add some additional debugging output for use re Debug => ALL to debug minlen/stopmin/delta issues.

It also does some code cleanup and commenting to make things easier to understand.

demerphq committed:
regcomp.c: enhance S_debug_studydata to show min/stopmin/delta

Also call it many more times during the study_chunk() process.

This is helpful for debuging minlen related issues.

Note this function is not in embed.fnc and is used strictly inside
of the regex engine, so no changes there.

demerphq committed:
regcomp.c: reorder and comment S_study_chunk() internal vars

The list was kinda random and did not include many comments, which made
life difficult understanding what the purpose of the different vars.

This documents them, and includes a follow up to investigate that came
up during the documentation process: first_non_open seems a bit off.
Will investigate and improve further in a subsequent patch.

demerphq committed:
regcomp.c: minor blank line removal/insertion for clarity

demerphq committed:
regcomp.c: deal with stopmin and min properly

stopmin is set when we encounter an ACCEPT, it basically says "even
though the minlen might look like X it is actually a smaller Y". It also
implies that delta (which refers to the max length a pattern might
match) should be at least a certain size. This was not being handled
properly nor propagated to callers in all situations. This in particular
affected use of ACCEPT inside of lookbehind. This also made final_minlen
redundant and it has been removed.

demerphq committed:
re_tests: ACCEPT followed by UNLESSM fixed width negative lookbehind

Make sure that things with END type opcodes don't mess up ACCEPT paren
close logic.

demerphq committed:
re_tests: ACCEPT inside of named capture accessed via GOSUB

demerphq committed:
re_tests: ACCEPT in IFMATCH: variable positive lookbehind

demerphq committed:
re_tests: ACCEPT in UNLESSM variable negative lookbehind

demerphq committed:
regcomp.c: ACCEPT inside of a (...)+ should disable mandatory substrings

GH Issue #19484 reported that

    print "ABDE" =~ /(A (A|B(*ACCEPT)|C)+ D)(E)/x ? "yes: <$1-$2>" : "no";

does not output the expected 'AB-B', and instead does not match.
Removing the + quantifier behaves as expected.

This patch is 1/4 and fixes part of the problem: the regex optimizer
S_study_chunk() was not handling the ACCEPT properly and assuming the
pattern MUST contain 'A' and 'DE', however this is wrong, the ACCEPT
means that it must only contain 'A' and the 'DE' is actually optional.

demerphq committed:
regexec.c: ACCEPT inside of a (...)+ should stop looping (CURLYM optimization)

GH Issue #19484 reported that

    print "ABDE" =~ /(A (A|B(*ACCEPT)|C)+ D)(E)/x ? "yes: <$1-$2>" : "no";

does not output the expected 'AB-B', and instead does not match.
Removing the + quantifier behaves as expected.

This patch is 3/4 and fixes part of the problem: the CURLYM optimization
was not terminating its loop properly when it contained an ACCEPT. This
patch adds a new variable 'is_accepted' which is used to ensure that the
CURLYM optimization stops after an ACCEPT regop is executed.

demerphq committed:
regexec.c: make ACCEPT close logic handle SUCCEED/LOOKBEHIND_END opcodes

GH Issue #19484 reported that

    print "ABDE" =~ /(A (A|B(*ACCEPT)|C)+ D)(E)/x ? "yes: <$1-$2>" : "no";

does not output the expected 'AB-B', and instead does not match.
Removing the + quantifier behaves as expected.

This patch is 4/4 of the patches to fix this problem: SUCCEED and
LOOKBEHIND_END regops are type 'END' which have a next_off of 0. This
was causing regnext() to return null inside of the loop iterator for the
logic in ACCEPT which closes any open capture buffers thus terminating
the loop prematurely and preventing some of the capture buffers from
being properly closed. SUCCEED is used to end a variety of structures,
including lookahead IFMATCH and UNLESSM, SUSPEND, and CURLYM, and
LOOKBEHIND_END serves the same purpose for lookbehind IFMATCH and
UNLESSM. Thus this patch fixes the original bug but also fixes a variety
of other cases involving ACCEPT.

demerphq committed:
re_tests: ACCEPT with CURLYM optimization

demerphq committed:
regexec.c: Fix up for ACCEPT inside of a (...)+ set lastopen in CURLYM

GH Issue #19484 reported that

    print "ABDE" =~ /(A (A|B(*ACCEPT)|C)+ D)(E)/x ? "yes: <$1-$2>" : "no";

does not output the expected 'AB-B', and instead does not match.
Removing the + quantifier behaves as expected.

This patch is 2/4 and fixes part of the problem: lastopen was not
being set properly inside of the CURLYM optimization. lastopen is used
by the ACCEPT logic to know which parens need to be closed.

demerphq committed:
re_tests: ACCEPT followed by SUSPEND

demerphq committed:
re_tests: ACCEPT followed by IFMATCH fixed width pos lookbehind

demerphq committed:
reg_mesg.t: check that ACCEPT in capturing variable length lookbehind warns

hvds commented on regcomp.c:
```diff
@@ -5715,6 +5715,16 @@ S_study_chunk(pTHX_
                 }
+
+                if (data && data->flags & SCF_SEEN_ACCEPT) {
+                    if (flags & SCF_DO_SUBSTR) {
```
Should we also check for `SCF_DO_STCLASS` here?

demerphq replied:
No, not as far as I can tell. Although while verifying this I found a problem caused by different code. The start class stuff should be fine as long as the min-length logic is correct. I will push an additional commit to deal with that.

demerphq replied:
So the reason (*ACCEPT) doesn't affect stclass (eg ANYOF style) in the optimizer in the same way it does substrings is because *ACCEPT affects the minimum length string that the pattern might match, but it doesn't affect the things it might start with. The only real effect it might have on startclass would be to allow it to accept the empty string. Put another way, (*ACCEPT) means that things to its right cant be mandatory, but they can still be something that starts a match. The only case where this is not the case is when ACCEPT would cause the pattern to match the empty and in that case there shouldnt be a startclass at all, and it should be handled via the minlen logic. This is why a couple of the later patches in this PR are there, to ensure that minlen is set properly.  BTW, those patches wouldn't exist if you hadn't asked this question so thanks!

Consider a case like this (using charclasses here to prevent this being turned into a mandatory substring question).

`/[aA]*[bB]/`

it should have a stclass of `ANYOF[aAbB]` and have a minlen of 1.

But now if we insert an `(*ACCEPT)` in the middle:

`/[aA]*(*ACCEPT)[bB]/`

it has a minlen of 0, and will match any string and the optimizer already understands that if the pattern can match the empty string then there is no startclass at all.

Change the * to a +, and the ACCEPT doesn't affect the stclass at all, it will be the same with or without the ACCEPT, any `ANYOF[aA]`.

Consider a case like `/ [aA]* ( [bB] | [cC](*ACCEPT) | [dD]*) [eE] /x`, this should have a stclass of `ANYOF[aAbBcCdDeE]`. The ACCEPT doesn't affect the fact that 'e' might be a legit start character for a matching string.

With mandatory substrings however the case is different. If I have `/a(?:b+(*ACCEPT))?c/` then the fixed string should be 'ab', c will not be part of a successful match. Compare with `/ab+c/`, where we have two mandatory strings to choose from 'ab' and 'bc'. 

With `/a* ( b | c | d ) e/` the mandatory substring would be 'e', but if i inject an accept into the alternation like this: `/a* (b | c (*ACCEPT) | d ) e/` then 'e' is now not mandatory, if it goes through the 'c' branch then 'e' will not be part of a match.

hvds replied:
Thanks, I'd overlooked the fact that minlen=0 makes stclass irrelevant: I guess that was implied by your previous comment, but I didn't pick up on it.

(I'd also recommend putting backticks round code fragments: one of your examples above is being parsed as URL markup in github's web interface, because it matches `[...](...)`.)

demerphq replied:
Good tip. I updated the comment, but ill keep it in mind for the future. :-)

hvds commented on regexec.c:
```diff
@@ -8975,7 +8978,7 @@ NULL
 
           curlym_do_B: /* execute the B in /A{m,n}B/  */
-            if (ST.Binfo.count < 0) {
+            if (!is_accepted && ST.Binfo.count < 0) {
```
Seems like after this block we can still execute the following diagnostic ("CURLYM trying tail") and the subsequent block (testing for "CURLYM Fast bail"), which I'm guessing we shouldn't if `is_accepted`.

demerphq replied:
Yes, right. Good catch.

hvds commented on regexec.c:
```diff
@@ -9019,21 +9022,22 @@ NULL
             if (ST.me->flags) {
```
Similarly, if `ST.me->flags` is false, we'll fall through to the `PUSH_STATE_GOTO`, which I'm guessing we shouldn't. (Guessing at all this stuff, took me quite a few minutes to dig out what `ST.me` was.)

demerphq replied:
Yes, also true, I will reorganize that logic a bit when i push.

hvds commented on regexec.c:
```diff
@@ -8397,21 +8398,29 @@ S_regmatch(pTHX_ regmatch_info *reginfo, char *startpos, regnode *prog)
+
+                    n = ARG(cursor);
+
+                    if ( n > lastopen )
+                        break;
```
If I read it correctly, this is an additional change to that described in the commit: previously it would have continued in this case. Can you clarify what it implies here if we see a `CLOSE(n)` with `n > lastopen`? Is making it break simply an additional optimization? At first glance it seems like this loop could correctly walk past `OPEN(lastopen+1) ... CLOSE(lastopen+1)` with a `CLOSE(lastopen)` still to come.

demerphq replied:
That should read "continue" not "break". Will be fixed in the repush.

hvds commented on regcomp.c:
```diff
@@ -6745,6 +6744,17 @@ S_study_chunk(pTHX_
+        and we need to ensure that the delta accounts for it. delta
+        represents the difference between min length and max length for
+        this part of the pattern. */
+        if (delta < min - stopmin)
+            delta = min - stopmin;
```
This is unexpected, I was expecting it to be an unconditional `delta += min - stopmin;`.

For `/ (A (*ACCEPT) | BC){2} D{0,4} /x` I think I'd expect before this block min=2 (for assumed match on AA), delta=6 (for match on BCBCDDDD), stopmin=1 (for match on A). After this block I'd expect min=1, delta=7, since the possible longest match has not changed.

demerphq replied:
Yes, concur. Will fix. It turns out it doesn't matter much, as the delta is only used for fixed string, and fixed string is disabled by accept. But you are correct in your analysis. While debugging the case you showed however I found a different issue with the {2}, but that is a different bug and will get its own commit.

hvds commented on t/re/re_tests:
```diff
@@ -2086,6 +2086,8 @@ AB\s+\x{100}	AB \x{100}X	y	-	-
 /(A(A|B(*ACCEPT)|(?>C))D)(E)/	ABCDE	y	$&	AB	-	# ACCEPT with SUSPEND in pattern
 /(A(A|B(*ACCEPT)|C(?<=C))D)(E)/	ABCDE	y	$1-$2	AB-B	-	# ACCEPT with IFMATCH in pattern
 /(A(A|B(*ACCEPT)|C(?<=C))D)(E)/	ABCDE	y	$&	AB	-	# ACCEPT with IFMATCH in pattern
+/(A(A|B(*ACCEPT)|C(?<!Q))D)(E)/	ABCDE	y	$1-$2	AB-B	-	# ACCEPT with UNLESSM in pattern
+/(A(A|B(*ACCEPT)|C(?<!Q))D)(E)/	ABCDE	y	$&	AB	-	# ACCEPT with UNLESSM in pattern
```
These don't match the commit message: the ACCEPT is not _inside_ UNLESSM.

demerphq replied:
Test renamed, we are checking that the LOOKBEHIND_END does not break accept capture close logic.

hvds reviewed (commented):
It may also be worth adding tests to `t/re/opt.t` to more specifically declare the expected minlen, maxlen etc that we expect to be discovered for these cases.

khwilliamson commented on regcomp.c:
```diff
@@ -4644,20 +4646,41 @@ S_study_chunk(pTHX_
-    SSize_t delta = 0;
+    /* vars about the regnodes we are working with */
+    regnode *scan = *scanp; /* the current opcode we are inspecting */
+    regnode *next = NULL;   /* the next opcode from scan, tmp var */
```
I think s/from/beyond/ is clearer

demerphq replied:
changed.

khwilliamson reviewed (commented):
It would take me a lot of time to familiarize myself with the optimizer in order to adequately review these changes.  Except for a trivial note about a comment, it all looks reasonable to me.  But I'm close to having zero knowledge about this area.

demerphq committed:
regcomp.c: With ACCEPT set stopmin even if no data struct present

Otherwise top level branches can end up with mistaken minlen with
ACCEPT. Eg, /A(*ACCEPT)B/ does not require B to be present.

demerphq committed:
regcomp.c: fix substring optimizer for ACCEPT inside of CURLY

ACCEPT essentially overrides quantifiers larger than 1. Eg,
/(A){2}/ has a mincount of 2 and a maxcount of 2, and a minlen
of 2 for "AA". But /(A(*ACCEPT)){2}/ should actually have a
mincount of 1, and a minlen of 1 as it can match 'A'. In the regex
engine proper this doesn't matter, we just do the right thing. But
in the optimizer it matters. This patch sets the mincount to 1 in
such cases whenever the contents contains an ACCEPT.

Thanks for Hugo for asking the questions that lead to this patch.

demerphq commented:
I did a force push with the changes rebased. I had to do a rebase anyway as there were conflicts, sorry about that Hugo. I have changed pretty much everything mentioned in feedback, and added a couple of extra commits that came up while digging into Hugo's questions.

demerphq commented:
@hvds waiting on you on this one mate.

demerphq commented:
I have added the release blocker tag to this PR because it fixes some bugs related to not throwing experimental warnings about variable look-behind with capturing. Something like this:
```
/(?<=(foo|bar(*ACCEPT)|baz)bop)/

is actually variable length, and can match a string that is either 3 or 6 characters long. Without this patch the optimizer thinks all of the options are 6 characters long and therefore does not produce experimental warnings like it should.

hvds commented on t/re/reg_mesg.t:
```diff
@@ -747,6 +747,12 @@ my @experimental_vlb = (
     '/(?| (?=(foo)) | (?<!(foo)|p) )/x'
             => 'Variable length negative lookbehind with capturing' .
                ' is experimental {#} m/(?| (?=(foo)) | (?<!(foo)|p) ){#}/',
+    '/(?<!(foo|bloop(*ACCEPT)|bar)baz)/'
+            => 'Variable length negative lookbehind with capturing' .
+               ' is experimental {#} m/(?<!(foo|bloop(*ACCEPT)|bar)baz){#}/',
+    '/(?<=(foo|bloop(*ACCEPT)|bar)baz)/'
+            => 'Variable length positive lookbehind with capturing' .
+               ' is experimental {#} m/(?<=(foo|bloop(*ACCEPT)|bar)baz){#}/',
```
I would have expected this to test something where the ACCEPT was what changed it from fixed to variable length: `/(?<=(ab|c(*ACCEPT)d))/`, and perhaps also the reverse: `/(?<=(ab|cd(*ACCEPT)efg))/` (though if the latter were wrongly to warn I'd have no problem living with it).

demerphq replied:
Ah good suggestion. Ill tweak that and push a fix now for rebase afterwards. (I'm learning. ;-)

demerphq replied:
Pushed.

demerphq replied:
BTW, I didn't add `/(?<=(ab|cd(*ACCEPT)efg))/` as it does warn as you suspected it might, and I want to add a new issue to make it not warn. But I don't want to burden this PR with that fix as it could be serious work and I'd really like to see this released in 5.36. I just had some ideas how to make this better, but perfect is the enemy of good enough and all that.

hvds replied:
> BTW, I didn't add `/(?<=(ab|cd(*ACCEPT)efg))/` as it does warn as you suspected it might, and I want to add a new issue to make it not warn. But I don't want to burden this PR with that fix as it could be serious work and I'd really like to see this released in 5.36. I just had some ideas how to make this better, but perfect is the enemy of good enough and all that.

Yep, that's fine by me.

hvds reviewed (approved):
LGTM, the one new comment is optional.

demerphq commented:
Based on @rjbs  mail that we should be able to merge fixes that are unlikely to cause problems to anything other than the feature they pertain to, which this patch is I am merging now. Thanks for the review @hvds and @khwilliamson.
